New Tests in Model Serving: Verify changes to existing model resources

manosnoam opened the pull request:
This change includes new tests and logic to verify the following use cases during Sanity tests:

- Regression for bugs RHOAIENG-549, RHOAIENG-2759, RHOAIENG-2724, RHOAIENG-2869.
- Adding both secured and regular model server in the same project.
- Changing existing model framework and retest endpoint.
- Adding new models in existing model server.
- Running multiple projects with the same model.

The suite reveals 3 of the mentioned bugs above, which are also reported as **Product Bugs** if there is no workaround:

![image](https://github.com/red-hat-data-services/ods-ci/assets/9913945/ff5b3faf-efe2-442f-8f3f-5c1140084f6e)

As part of this PR many improvements were added:

- Run "Clean All" only in Suite Setup and Teardown (but not as part of the tests as it was before).
- Optimize timeouts in few Model Serve Keywords.
- KW to switch projects in Model serving to deal with same model name under "All Projects".
- Delete Model via UI according to model name inside its parent project.
- Create Data Science Project KW option to skip if project already exists.
- Add KW 'Recreate S3 Data Connection' to edit existing connection, or create new if doesn't exist.
- Add option to update existing server in 'Create Model Server' Keyword.
- Add option to update existing model in 'Serve Model' Keyword.
- Add option in 'Navigate To Page' KW to click items without sub-menu.
- Remove duplicated Keywords.
- Rename "Try Opening Create Server" to "Clean All Models Of Current User".
- KW to 'Get Modelmesh Events And Logs' by specific server and project.

![image](https://github.com/red-hat-data-services/ods-ci/assets/9913945/f9215db0-672c-458b-822b-0bbbe5cec9cd)

bdattoma commented on ods_ci/tests/Resources/Page/ODH/ODHDashboard/ODHDataScienceProject/ModelServer.resource:
```diff
@@ -32,7 +32,12 @@ Create Model Server
+    IF    ${existing_server} == ${TRUE}
+        ${is_exist} =  Run Keyword And Return Status     Wait Until Page Contains Element    //button[.="${server_name}"]
+        IF  ${is_exist}    Run Keyword And Return
+        ...    Log    Model Server '${server_name}' already exists, reusing server    console=True
+    END
     SeleniumLibrary.Click Button    Add model server
```
it is not re-using the server if it clicks on "Add model server"

bdattoma commented on ods_ci/tests/Resources/Page/ODH/ODHDashboard/ODHDataScienceProject/ModelServer.resource:
```diff
@@ -59,7 +64,8 @@ Create Model Server
     ELSE IF    ${token}==${TRUE}
         Enable Token Authentication
     END
-    Click Button    Add
```
have you seen this failing without the wait?

manosnoam replied:
yes, it happened. That's why making sure the the "Add" button is enabled before pressing it.

bdattoma commented on ods_ci/tests/Resources/Page/ODH/ODHDashboard/ODHDataScienceProject/DataConnections.resource:
```diff
@@ -160,3 +160,26 @@ Get Data Connection Form Values
     ${s3_bucket}   get element attribute    ${S3_BUCKET_DC_INPUT_XP}      value
     [return]    ${s3_name}    ${s3_key}    ${s3_secret}    ${s3_endpoint}    ${s3_region}    ${s3_bucket}
+
+Recreate S3 Data Connection
```
maybe the name is not very intuitive. It sounds like "delete the data connection and create it again", but it's not what id does

bdattoma commented on ods_ci/tests/Resources/Page/ODH/ODHDashboard/ODHDataScienceProject/DataConnections.resource:
```diff
@@ -160,3 +160,26 @@ Get Data Connection Form Values
+    ${is_exist} =  Run Keyword And Return Status
+        ...    Wait Until Page Contains Element    ${DC_SECTION_XP}//tr[td[@data-label="Name"]/*/div[text()="${dc_name}"]]
+    IF  "${is_exist}" == "${TRUE}"   
+        Log    Data Connection '${dc_name}' already exists, reusing connection    console=True
```
it is not re-using but it's editing, I think it's different and should be clear when using the kw

bdattoma commented on ods_ci/tests/Tests/400__ods_dashboard/420__model_serving/421__model_serving_gpu.robot:
```diff
@@ -117,25 +117,6 @@ Clean Up DSP Page
         Wait Until Page Contains    Data Science Projects
     END
 
-Try Opening Create Server
```
don't you need to delete also `Clean Up DSP Page`?

manosnoam replied:
Right!

bdattoma commented on ods_ci/tests/Resources/Page/ODH/ODHDashboard/ODHDataScienceProject/Projects.resource:
```diff
@@ -110,22 +110,31 @@ Project's Owner Should Be
+    END
+    IF  "${is_exist}" == "${TRUE}"   
+        Log    DS Project '${title}' already exists, reusing project    console=True
+        Click Link      ${title}
```
of course it works, it wasn't the point of the comment

I wonder how much longer it would actually take. But ok

manosnoam commented on ods_ci/tests/Resources/Page/ODH/ODHDashboard/ODHModelServing.resource:
```diff
@@ -31,7 +31,7 @@ ${LLM_RESOURCES_DIRPATH}=    ods_ci/tests/Resources/Files/llm
 Open Model Serving Home Page
     [Documentation]    Verifies submenu Settings > "Model Serving" is visible and navigates to page
-    SeleniumLibrary.Page Should Contain    Model Serving
+    Wait Until Page Contains Element    xpath://a[text()='Model Serving']
```
yes, it happened that previous action (e.g. creating Model server) finished but the link to the "Model Serving" link was not yet available. That's why making sure before clicking it.

bdattoma replied:
the Model Serving link is not correlated to creating model server action

manosnoam replied:
ack, I added a new option with `Navigate To Page` to get to to Model Serving.

manosnoam commented on ods_ci/tests/Resources/Page/ODH/ODHDashboard/ODHModelServing.resource:
```diff
@@ -52,12 +52,22 @@ Serve Model
+    IF  "${is_exist}" == "${TRUE}"
+        Log    Model '${model_name}' already exists, editing model   console=True
+        Click Element    ${model_row_xpath}${MS_TABLE_DROPDOWN_TOGGLE}
+        Click Element    ${model_row_xpath}${MS_TABLE_DROPDOWN_TOGGLE}/..//button[.="Edit"]
```
Will try!

manosnoam replied:
`ODHDashboard.Click Action From Actions Menu` does it, thanks!

manosnoam commented on ods_ci/tests/Tests/400__ods_dashboard/420__model_serving/420__model_serving.robot:
```diff
@@ -105,29 +69,88 @@ Verify Tensorflow Model Via UI
+    [Documentation]    Test the deployment of multiple DS project with same openvino_ir model
+    [Tags]    Sanity
+    ...    RHOAIENG-549    RHOAIENG-2724
+    FOR  ${idx}  IN RANGE  1  6
```
ack, will use an internal keyword

manosnoam replied:
Done

manosnoam commented on ods_ci/tests/Tests/400__ods_dashboard/420__model_serving/420__model_serving.robot:
```diff
@@ -35,65 +37,27 @@ Verify Model Serving Installation
     Wait for RHODS Dashboard to Load    wait_for_cards=${FALSE}    expected_page=Data Science Projects
-    Create Data Science Project    title=${PRJ_TITLE}    description=${PRJ_DESCRIPTION}
-    Create S3 Data Connection    project_title=${PRJ_TITLE}    dc_name=model-serving-connection
+    Create Data Science Project    title=${PRJ_TITLE}    description=${PRJ_DESCRIPTION}    existing_project=${TRUE}
```
`Create Data Science Project If Not Exists` requires a username. 

The option added to `Create Data Science Project` is a simple flag where we can directly use `Create Data Science Project` to check if the project already exists, and if it is - use it, but if it is not - create it. 
The flag is similar to what we already have in other keywords, which is set to be false (do not use existing) by default, i.e. `existing_resource=${FALSE}`.

manosnoam commented on ods_ci/tests/Tests/400__ods_dashboard/420__model_serving/LLMs/422__model_serving_llm_other_runtimes_UI.robot:
```diff
@@ -50,7 +50,7 @@ Verify Non Admin Can Serve And Query A Model Using The UI  # robocop: disable
     ...    thanos_url=${THANOS_URL}    thanos_token=${THANOS_TOKEN}
     ...    model_name=${model_name}    query_kind=single    namespace=${test_namespace}    period=5m    exp_value=1
-    Delete Model Via UI    ${model_name}
+    Delete Model Via UI    ${model_name}    ${test_namespace}
```
Thanks for pointing out! I fixed it so it just delete by model name, without project name.
Also re-tested:
![image](https://github.com/red-hat-data-services/ods-ci/assets/9913945/e6851c5c-8d90-45f7-b6cb-7d558aed7015)

manosnoam commented on ods_ci/tests/Tests/400__ods_dashboard/420__model_serving/LLMs/422__model_serving_llm_UI.robot:
```diff
@@ -319,7 +319,7 @@ Verify User With Admin Permission Can Deploy Query And Delete A LLM
     ...    inference_type=streaming    n_times=1
     ...    namespace=${test_namespace}    protocol=http    validate_response=${FALSE}
-    Delete Model Via UI    ${flan_model_name}
+    Delete Model Via UI    ${flan_model_name}    ${test_namespace}
```
Fixed: https://github.com/red-hat-data-services/ods-ci/pull/1185#discussion_r1483566190

manosnoam commented:
Have verified the whole suite many times. 
Running it has revealed few bugs, as seen bellow (and reported):
![image](https://github.com/red-hat-data-services/ods-ci/assets/9913945/fb068ef4-e92e-4449-ba9c-66fd2ee4e3d5)

bdattoma commented:
> Have verified the whole suite many times. Running it has revealed few bugs, as seen bellow (and reported): ![image](https://private-user-images.githubusercontent.com/9913945/303517283-fb068ef4-e92e-4449-ba9c-66fd2ee4e3d5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc0NzA3MDEsIm5iZiI6MTcwNzQ3MDQwMSwicGF0aCI6Ii85OTEzOTQ1LzMwMzUxNzI4My1mYjA2OGVmNC1lOTJlLTQ0NDktYmE5Yy02NmZkMmVlNGUzZDUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMDlUMDkyMDAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWQ0NjQzM2MxOGU0MjVhNjk4Njk3ZjVhZjEwZDRmNjY0NWNiY2Y5OWRmMWVlYTIwNjgyNjIwYzExN2M1ZmNjNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.f6D2lEiO2q3bRDwghiPx6E6Z210fP9jYIQ6-kAL1YT4)

Please do also regresstion tests with tests outside this suite (e.g., LLMs)

manosnoam commented:
> Please do also regresstion tests with tests outside this suite (e.g., LLMs)
Verified few times with other LLM tests, for example:
![image](https://github.com/red-hat-data-services/ods-ci/assets/9913945/4fc75b8a-d25f-4fe4-aa72-e64009d3373d)

> Also, pls check the dry run, there is a failure
The dry-run most likely failed for you after I rebased (complected merge). I fixed it.
